# Evaluates text submissions.

This code defines a scoring system for a competition that evaluates text submissions. It calculates the mean perplexity of submitted text permutations compared to original texts using a pre-trained language model. It ensures that the submissions are valid permutations and uses a custom PerplexityCalculator class to compute perplexity scores. Lower scores indicate better quality submissions.


In [3]:
import gc
import os
from math import exp
from collections import Counter
from typing import List, Optional, Union

import numpy as np
import pandas as pd
import transformers
import torch

os.environ['OMP_NUM_THREADS'] = '1'
os.environ['TOKENIZERS_PARALLELISM'] = 'false'
PAD_TOKEN_LABEL_ID = torch.nn.CrossEntropyLoss().ignore_index
DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')


class ParticipantVisibleError(Exception):
    pass


def score(
    solution: pd.DataFrame,
    submission: pd.DataFrame,
    row_id_column_name: str,
    model_path: str = '/kaggle/input/gemma-2/transformers/gemma-2-9b/2',
    load_in_8bit: bool = True,
    clear_mem: bool = False,
) -> float:
    """
    Calculates the mean perplexity of submitted text permutations compared to an original text.

    Parameters
    ----------
    solution : DataFrame
        DataFrame containing the original text in a column named 'text'.
        Includes a row ID column specified by `row_id_column_name`.

    submission : DataFrame
        DataFrame containing the permuted text in a column named 'text'.
        Must have the same row IDs as the solution.
        Includes a row ID column specified by `row_id_column_name`.

    row_id_column_name : str
        Name of the column containing row IDs.
        Ensures aligned comparison between solution and submission.

    model_path : str
        Path to the serialized LLM.

    clear_mem : bool
        Clear GPU memory after scoring by clearing the CUDA cache.
        Useful for testing.

    Returns
    -------
    float
        The mean perplexity score. Lower is better.

    Raises
    ------
    ParticipantVisibleError
        If the submission format is invalid or submitted strings are not valid permutations.

    Examples
    --------
    >>> import pandas as pd
    >>> model_path = "/kaggle/input/gemma-2/transformers/gemma-2-9b/2"
    >>> solution = pd.DataFrame({
    ...     'id': [0, 1],
    ...     'text': ["this is a normal english sentence", "the quick brown fox jumps over the lazy dog"]
    ... })
    >>> submission = pd.DataFrame({
    ...     'id': [0, 1],
    ...     'text': ["sentence english normal a is this", "lazy the over jumps fox brown quick the dog"]
    ... })
    >>> score(solution, submission, 'id', model_path=model_path, clear_mem=True) > 0
    True
    """
    # Check that each submitted string is a permutation of the solution string
    sol_counts = solution.loc[:, 'text'].str.split().apply(Counter)
    sub_counts = submission.loc[:, 'text'].str.split().apply(Counter)
    invalid_mask = sol_counts != sub_counts
    if invalid_mask.any():
        raise ParticipantVisibleError(
            'At least one submitted string is not a valid permutation of the solution string.'
        )

    # Calculate perplexity for the submitted strings
    sub_strings = [
        ' '.join(s.split()) for s in submission['text'].tolist()
    ]  # Split and rejoin to normalize whitespace
    scorer = PerplexityCalculator(
        model_path=model_path,
        load_in_8bit=load_in_8bit,
    )  # Initialize the perplexity calculator with a pre-trained model
    perplexities = scorer.get_perplexity(
        sub_strings
    )  # Calculate perplexity for each submitted string

    if clear_mem:
        # Just move on if it fails. Not essential if we have the score.
        try:
            scorer.clear_gpu_memory()
        except:
            print('GPU memory clearing failed.')

    return float(np.mean(perplexities))


class PerplexityCalculator:
    """
    Calculates perplexity of text using a pre-trained language model.

    Adapted from https://github.com/asahi417/lmppl/blob/main/lmppl/ppl_recurrent_lm.py

    Parameters
    ----------
    model_path : str
        Path to the pre-trained language model

    load_in_8bit : bool, default=False
        Use 8-bit quantization for the model. Requires CUDA.

    device_map : str, default="auto"
        Device mapping for the model.
    """

    def __init__(
        self,
        model_path: str,
        load_in_8bit: bool = False,
        device_map: str = 'auto',
    ):
        self.tokenizer = transformers.AutoTokenizer.from_pretrained(model_path,padding_side="right")
        # Configure model loading based on quantization setting and device availability
        if load_in_8bit:
            if DEVICE.type != 'cuda':
                raise ValueError('8-bit quantization requires CUDA device')
                
            #quantization_config = transformers.BitsAndBytesConfig(load_in_8bit=True)
            #quantization_config = transformers.BitsAndBytesConfig(load_in_4bit=True)

            quantization_config = transformers.BitsAndBytesConfig(
                load_in_4bit = True,
                bnb_4bit_quant_type = "fp4", #fp4 nf4
                bnb_4bit_use_double_quant = False,
                bnb_4bit_compute_dtype=torch.float16,
            )
            
            self.model = transformers.AutoModelForCausalLM.from_pretrained(
                model_path,
                quantization_config=quantization_config,
                device_map=device_map,
            )
        else:
            self.model = transformers.AutoModelForCausalLM.from_pretrained(
                model_path,
                torch_dtype=torch.float16 if DEVICE.type == 'cuda' else torch.float32,
                device_map=device_map,
            )

        self.loss_fct = torch.nn.CrossEntropyLoss(reduction='none')

        self.model.eval()
        #if not load_in_8bit:
        #    self.model.to(DEVICE)  # Explicitly move the model to the device

    def get_perplexity(
        self, input_texts: Union[str, List[str]], batch_size: 32
    ) -> Union[float, List[float]]:
        """
        Calculates the perplexity of given texts.

        Parameters
        ----------
        input_texts : str or list of str
            A single string or a list of strings.

        batch_size : int, default=None
            Batch size for processing. Defaults to the number of input texts.

        verbose : bool, default=False
            Display progress bar.

        Returns
        -------
        float or list of float
            A single perplexity value if input is a single string,
            or a list of perplexity values if input is a list of strings.

        Examples
        --------
        >>> import pandas as pd
        >>> model_path = "/kaggle/input/gemma-2/transformers/gemma-2-9b/2"
        >>> scorer = PerplexityCalculator(model_path=model_path)

        >>> submission = pd.DataFrame({
        ...     'id': [0, 1, 2],
        ...     'text': ["this is a normal english sentence", "thsi is a slihgtly misspelled zr4g sentense", "the quick brown fox jumps over the lazy dog"]
        ... })
        >>> perplexities = scorer.get_perplexity(submission["text"].tolist())
        >>> perplexities[0] < perplexities[1]
        True
        >>> perplexities[2] < perplexities[0]
        True

        >>> perplexities = scorer.get_perplexity(["this is a sentence", "another sentence"])
        >>> all(p > 0 for p in perplexities)
        True

        >>> scorer.clear_gpu_memory()
        """
        single_input = isinstance(input_texts, str)
        input_texts = [input_texts] if single_input else input_texts

        loss_list = []

        batches = len(input_texts)//batch_size + (len(input_texts)%batch_size != 0)
        for j in range(batches):
            
            a = j*batch_size
            b = (j+1)*batch_size
            input_batch = input_texts[a:b]
        
            with torch.no_grad():

                # Explicitly add sequence boundary tokens to the text
                text_with_special = [f"{self.tokenizer.bos_token}{text}{self.tokenizer.eos_token}" for text in input_batch]

                # Tokenize
                model_inputs = self.tokenizer(
                    text_with_special,
                    return_tensors='pt',
                    add_special_tokens=False,
                    padding=True
                )

                if 'token_type_ids' in model_inputs:
                    model_inputs.pop('token_type_ids')

                model_inputs = {k: v.to(DEVICE) for k, v in model_inputs.items()}

                # Get model output
                output = self.model(**model_inputs, use_cache=False)
                logits = output['logits']

                label = model_inputs['input_ids']
                label[label == self.tokenizer.pad_token_id] = PAD_TOKEN_LABEL_ID

                # Shift logits and labels for calculating loss
                shift_logits = logits[..., :-1, :].contiguous()  # Drop last prediction
                shift_labels = label[..., 1:].contiguous()  # Drop first input

                # Calculate token-wise loss
                loss = self.loss_fct(
                    shift_logits.view(-1, shift_logits.size(-1)),
                    shift_labels.view(-1)
                )

                loss = loss.view(len(logits), -1)
                valid_length = (shift_labels != PAD_TOKEN_LABEL_ID).sum(dim=-1)
                loss = torch.sum(loss, -1) / valid_length

                loss_list += loss.cpu().tolist()

                # Debug output
                #print(f"\nProcessing: '{text}'")
                #print(f"With special tokens: '{text_with_special}'")
                #print(f"Input tokens: {model_inputs['input_ids'][0].tolist()}")
                #print(f"Target tokens: {shift_labels[0].tolist()}")
                #print(f"Input decoded: {self.tokenizer.decode(model_inputs['input_ids'][0])}")
                #print(f"Target decoded: {self.tokenizer.decode(shift_labels[0])}")
                #print(f"Individual losses: {loss.tolist()}")
                #print(f"Average loss: {sequence_loss.item():.4f}")

        ppl = [exp(i) for i in loss_list]

        # print("\nFinal perplexities:")
        # for text, perp in zip(input_texts, ppl):
        #     print(f"Text: '{text}'")
        #     print(f"Perplexity: {perp:.2f}")

        return ppl[0] if single_input else ppl

    def clear_gpu_memory(self) -> None:
        """Clears GPU memory by deleting references and emptying caches."""
        if not torch.cuda.is_available():
            return

        # Delete model and tokenizer if they exist
        if hasattr(self, 'model'):
            del self.model
        if hasattr(self, 'tokenizer'):
            del self.tokenizer

        # Run garbage collection
        gc.collect()

        # Clear CUDA cache and reset memory stats
        with DEVICE:
            torch.cuda.empty_cache()
            torch.cuda.ipc_collect()
            torch.cuda.reset_peak_memory_stats()

In [3]:
import itertools, math

df = pd.read_csv("/kaggle/input/santa-2024/sample_submission.csv")
df


id                                               text
0   0  advent chimney elf family fireplace gingerbrea...
1   1  advent chimney elf family fireplace gingerbrea...
2   2  yuletide decorations gifts cheer holiday carol...

# implementation of Ant Colony Optimization
This code implements an Ant Colony Optimization (ACO) algorithm for optimizing word permutations based on perplexity scores using a language model. It iteratively finds the best permutation of words to minimize perplexity, incorporates early stopping based on performance, and includes a solver function for processing text data for competitions like the Santa 2024 Perplexity Permutation Puzzle.

In [9]:
import numpy as np
import pandas as pd
import torch
import itertools
import random

class PerplexityPermutationACO:
    def __init__(
        self, 
        scorer, 
        original_text, 
        n_ants=50, 
        n_iterations=100, 
        alpha=1.0, 
        beta=2.0, 
        evaporation_rate=0.5,
        batch_size=32,
        patience=6  # New parameter for early stopping
    ):
        """
        Initialize the Ant Colony Optimization algorithm for permutation optimization.
        
        Parameters:
        -----------
        scorer : PerplexityCalculator
            The perplexity scoring model
        original_text : str
            The original text to be permuted
        n_ants : int
            Number of ants (solutions) in each iteration
        n_iterations : int
            Maximum number of iterations to run the optimization
        alpha : float
            Pheromone importance factor
        beta : float
            Heuristic importance factor
        evaporation_rate : float
            Rate of pheromone evaporation
        batch_size : int
            Batch size for perplexity calculation
        patience : int
            Number of iterations to wait for improvement before stopping
        """
        # Add patience to instance attributes
        self.patience = patience
        self.batch_size = batch_size

        self.words = original_text.split()
        self.n_words = len(self.words)
        self.scorer = scorer
        
        # Optimization parameters
        self.n_ants = n_ants
        self.n_iterations = n_iterations
        self.alpha = alpha
        self.beta = beta
        self.evaporation_rate = evaporation_rate
        
        # Initialize pheromone matrix
        self.pheromone = np.ones((self.n_words, self.n_words)) / self.n_words
        
        # Best solution tracking
        self.best_permutation = None
        self.best_perplexity = float('inf')
    
    def _construct_solution(self):
        """
        Construct a solution using probabilistic selection based on pheromones.
        
        Returns:
        --------
        list
            A permuted list of words
        """
        solution = []
        available_indices = list(range(self.n_words))
        
        while available_indices:
            # Compute selection probabilities
            probabilities = np.zeros(len(available_indices))
            for j, idx in enumerate(available_indices):
                # Compute probability using pheromone and heuristic information
                pheromone_factor = self.pheromone[len(solution), idx] ** self.alpha
                probabilities[j] = pheromone_factor
            
            # Normalize probabilities
            probabilities /= probabilities.sum()
            
            # Select word probabilistically
            selected_index = np.random.choice(len(available_indices), p=probabilities)
            selected_word_index = available_indices.pop(selected_index)
            solution.append(self.words[selected_word_index])
        
        return solution
    
    def _update_pheromones(self, solutions, perplexities):
        """
        Update pheromone levels based on solution quality.
        
        Parameters:
        -----------
        solutions : list of lists
            Generated word permutations
        perplexities : list
            Corresponding perplexity scores
        """
        # Decay existing pheromones
        self.pheromone *= (1 - self.evaporation_rate)
        
        # Sort solutions by perplexity
        sorted_solutions = sorted(zip(solutions, perplexities), key=lambda x: x[1])
        
        # Deposit more pheromone for better solutions
        for (solution, perplexity), quality_factor in zip(
            sorted_solutions, 
            np.linspace(1, 0.1, len(sorted_solutions))
        ):
            for i, word_index in enumerate(solution):
                original_index = self.words.index(word_index)
                self.pheromone[i, original_index] += quality_factor / perplexity
    
    def optimize(self):
        """
        Run the Ant Colony Optimization algorithm with early stopping.
        
        Returns:
        --------
        tuple
            Best permutation and its perplexity
        """
        # Track iterations without improvement
        iterations_without_improvement = 0
        previous_best_perplexity = float('inf')
        
        for iteration in range(self.n_iterations):
            # Generate solutions
            solutions = []
            perplexities = []
            
            for _ in range(self.n_ants):
                solution = self._construct_solution()
                solution_text = " ".join(solution)
                
                # Use batch_size parameter when calling get_perplexity
                perplexity = self.scorer.get_perplexity(
                    solution_text, 
                    batch_size=self.batch_size
                )
                
                solutions.append(solution)
                perplexities.append(perplexity)
                
                # Update global best solution
                if perplexity < self.best_perplexity:
                    self.best_permutation = solution
                    self.best_perplexity = perplexity
            
            # Update pheromone levels
            self._update_pheromones(solutions, perplexities)
            
            # Check for improvement
            if self.best_perplexity < previous_best_perplexity:
                iterations_without_improvement = 0
                previous_best_perplexity = self.best_perplexity
            else:
                iterations_without_improvement += 1
            
            # Optional: print progress
            print(f"Iteration {iteration+1}/{self.n_iterations}: "
                  f"Best Perplexity = {self.best_perplexity:.4f}")
            
            # Early stopping condition
            if iterations_without_improvement >= self.patience:
                print(f"Early stopping triggered after {iteration+1} iterations.")
                break
        
        return self.best_permutation, self.best_perplexity

def solve_santa_2024_permutation(
    sample_submission_path, 
    model_path='/kaggle/input/gemma-2/transformers/gemma-2-9b/2',
    output_path='submission.csv'
):
    """
    Solve the Santa 2024 Perplexity Permutation Puzzle using ACO.
    
    Parameters:
    -----------
    sample_submission_path : str
        Path to the sample submission CSV
    model_path : str
        Path to the Gemma model
    output_path : str
        Path to save the output submission
    """
    # Initialize scorer
    from transformers import AutoTokenizer, AutoModelForCausalLM
    
    scorer = PerplexityCalculator(model_path)
    
    # Load sample submission
    df = pd.read_csv(sample_submission_path)
    
    # Process each row
    results = []
    for idx, row in df.iterrows():
        original_text = row['text']
        
        # Initialize and run ACO with early stopping
        aco = PerplexityPermutationACO(
            scorer=scorer, 
            original_text=original_text,
            n_ants=50,
            n_iterations=100,
            batch_size=32,
            patience=10  # Add patience for early stopping
        )
        
        best_permutation, best_perplexity = aco.optimize()
        
        # Store result
        results.append({
            'id': idx,
            'text': ' '.join(best_permutation)
        })
        
        print(f"Row {idx}: Best Perplexity = {best_perplexity:.4f}")
        
        # Optional: free up GPU memory after each iteration
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
    
    # Create submission DataFrame
    submission_df = pd.DataFrame(results)
    submission_df.to_csv(output_path, index=False)
    print(f"Submission saved to {output_path}")
    
    return submission_df

In [10]:
# Example usage
solve_santa_2024_permutation('/kaggle/input/santa-2024/sample_submission.csv')

Loading checkpoint shards:   0%|          | 0/8 [00:00<?, ?it/s]

Starting from v4.46, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)


Iteration 1/100: Best Perplexity = 1230.2282
Iteration 2/100: Best Perplexity = 1035.5767
Iteration 3/100: Best Perplexity = 965.4243
Iteration 4/100: Best Perplexity = 965.4243
Iteration 5/100: Best Perplexity = 965.4243
Iteration 6/100: Best Perplexity = 965.4243
Iteration 7/100: Best Perplexity = 965.4243
Iteration 8/100: Best Perplexity = 837.7315
Iteration 9/100: Best Perplexity = 626.9516
Iteration 10/100: Best Perplexity = 626.9516
Iteration 11/100: Best Perplexity = 626.9516
Iteration 12/100: Best Perplexity = 626.9516
Iteration 13/100: Best Perplexity = 626.9516
Iteration 14/100: Best Perplexity = 626.9516
Iteration 15/100: Best Perplexity = 618.5973
Iteration 16/100: Best Perplexity = 578.3719
Iteration 17/100: Best Perplexity = 578.3719
Iteration 18/100: Best Perplexity = 578.3719
Iteration 19/100: Best Perplexity = 578.3719
Iteration 20/100: Best Perplexity = 559.0029
Iteration 21/100: Best Perplexity = 559.0029
Iteration 22/100: Best Perplexity = 534.1312
Iteration 23/100:

id                                               text
0   0  reindeer mistletoe scrooge elf gingerbread orn...
1   1  reindeer mistletoe scrooge gingerbread advent ...
2   2  sleigh yuletide carol holly polar workshop gri...
3   3  sleigh yuletide jingle sing of stocking naught...
4   4  wreath believe paper it not merry have star yo...
5   5  grinch unwrap card greeting eggnog the night e...

In [1]:
534.1312+1042.7253+545.5997+740.9372+658.6463+441.4128

3963.4525

In [2]:
3963.4525/6

660.5754166666667

# Result
The implemented Ant Colony Optimization (ACO) algorithm achieved a mean perplexity score of 660.575 across all processed rows, demonstrating its capability to optimize word permutations effectively for minimizing perplexity. To further enhance this performance, 

We integrate Simulated Annealing, providing an additional layer of optimization to refine the results and potentially achieve better perplexity scores.

# Hybrid Optimization Algorithm: Simulated Annealing (SA) and Ant Colony Optimization (ACO)

This code implements a **Hybrid Optimization Algorithm** combining **Simulated Annealing (SA)** and **Ant Colony Optimization (ACO)** for solving permutation-based problems with a focus on optimizing perplexity, which is commonly used as a metric in language models. The hybrid approach leverages the strengths of both algorithms to improve the solution quality.

---

## **Key Features and Explanation**

### **1. Hybrid Framework**
The optimizer combines:
- **Simulated Annealing (SA):**
    - Explores the solution space through probabilistic acceptance of worse solutions to escape local minima.
    - Gradually cools down the "temperature," reducing the likelihood of accepting worse solutions over time.
- **ACO-Inspired Local Search:**
    - Mimics the behavior of ants finding optimal paths using pheromone trails.
    - Uses heuristics to refine solutions iteratively.

---

### **2. Parameters**

#### **SA Parameters:**
- **initial_temperature**: The starting temperature for SA.
- **cooling_rate**: The rate at which the temperature decreases.

#### **ACO Parameters:**
- **n_ants**: Number of solutions (ants) evaluated per iteration.
- **alpha** and **beta**: Weighting for pheromone importance versus heuristic factors.

#### **Optimization Settings:**
- **n_iterations**: Maximum number of iterations.
- **batch_size**: Adjusted for perplexity calculation speed.
- **early_stopping** and **patience**: Stop the algorithm early if no improvement is observed.


In [4]:
import numpy as np
import pandas as pd
import torch
import random
from typing import List, Tuple

class HybridPerplexityPermutationOptimizer:
    def __init__(
        self, 
        scorer, 
        original_text: str, 
        n_ants=30,               # Augmentation du nombre de fourmis
        n_iterations=200,         # Plus d'itérations
        alpha=0.5,                # Influence phéromones
        beta=1,                 # Influence heuristiques
        initial_temperature=30.0, # Température initiale réduite
        cooling_rate=0.98,        # Refroidissement plus lent
        batch_size=16,            # Réduction de la taille du batch pour accélérer
        patience=10,              # Augmentation de la patience
        early_stopping=True
    ):
        """
        Hybrid Ant Colony Optimization and Simulated Annealing for Permutation Optimization
        
        Parameters explained in previous implementations
        """
        self.words = original_text.split()
        self.n_words = len(self.words)
        self.scorer = scorer
        
        # SA parameters
        self.initial_temperature = initial_temperature
        self.cooling_rate = cooling_rate
        
        # ACO parameters
        self.n_ants = n_ants
        self.n_iterations = n_iterations
        self.alpha = alpha
        self.beta = beta
        self.batch_size = batch_size
        self.patience = patience
        self.early_stopping = early_stopping
        
        # Pheromone matrix for ACO
        self.pheromone = np.ones((self.n_words, self.n_words)) / self.n_words
        
        # Best solution tracking
        self.best_permutation = None
        self.best_perplexity = float('inf')
        self.no_improvement_counter = 0
    
    def _evaluate_permutation(self, permutation: List[str]) -> float:
        """Compute perplexity for a single permutation"""
        text = " ".join(permutation)
        return self.scorer.get_perplexity(text, batch_size=self.batch_size)
    
    def _initialize_permutation(self) -> List[str]:
        """Initialize a random permutation"""
        return random.sample(self.words, len(self.words))
    
    def _neighbor(self, permutation: List[str]) -> List[str]:
        """Generate a neighboring solution by swapping two elements"""
        neighbor = permutation.copy()
        i, j = random.sample(range(len(permutation)), 2)
        neighbor[i], neighbor[j] = neighbor[j], neighbor[i]
        return neighbor
    
    def _simulated_annealing(self) -> Tuple[List[str], float]:
        """Simulated Annealing Optimization"""
        current_solution = self._initialize_permutation()
        current_perplexity = self._evaluate_permutation(current_solution)
        best_solution = current_solution
        best_perplexity = current_perplexity
        
        temperature = self.initial_temperature
        
        while temperature > 1e-3:
            # Generate neighbor
            neighbor_solution = self._neighbor(current_solution)
            neighbor_perplexity = self._evaluate_permutation(neighbor_solution)
            
            # Acceptance probability
            delta = neighbor_perplexity - current_perplexity
            if delta < 0 or random.random() < np.exp(-delta / temperature):
                current_solution = neighbor_solution
                current_perplexity = neighbor_perplexity
            
            # Update best solution
            if current_perplexity < best_perplexity:
                best_solution = current_solution
                best_perplexity = current_perplexity
            
            # Cool down
            temperature *= self.cooling_rate
        
        return best_solution, best_perplexity

    def _aco_local_search(self, permutation: List[str], perplexity: float) -> Tuple[List[str], float]:
        """ACO-inspired local search"""
        best_permutation = permutation
        best_perplexity = perplexity
        for _ in range(3):  # Local search attempts
            candidate = self._neighbor(permutation)
            candidate_perplexity = self._evaluate_permutation(candidate)
            if candidate_perplexity < best_perplexity:
                best_permutation = candidate
                best_perplexity = candidate_perplexity
        return best_permutation, best_perplexity

    def optimize(self) -> Tuple[List[str], float]:
        """Hybrid ACO-SA Optimization Process"""
        best_solution, best_perplexity = self._simulated_annealing()
        for iteration in range(self.n_iterations):
            # ACO-inspired local search
            best_solution, best_perplexity = self._aco_local_search(best_solution, best_perplexity)
            print(f"Iteration {iteration+1}/{self.n_iterations}: Best Perplexity = {best_perplexity:.4f}")
            
            # Early stopping check
            if self.early_stopping:
                if best_perplexity < self.best_perplexity:
                    self.best_perplexity = best_perplexity
                    self.no_improvement_counter = 0
                else:
                    self.no_improvement_counter += 1
                    if self.no_improvement_counter >= self.patience:
                        print("Early stopping triggered.")
                        break
        
        return best_solution, best_perplexity


def solve_santa_2024_permutation_SA(
    sample_submission_path: str, 
    model_path: str = '/kaggle/input/gemma-2/transformers/gemma-2-9b/2',
    output_path: str = 'submission.csv'
):
    """
    Solve Santa 2024 Perplexity Permutation Puzzle using Hybrid ACO-SA
    """
    from transformers import AutoTokenizer, AutoModelForCausalLM
    
    # Initialize scorer
    scorer = PerplexityCalculator(model_path)
    
    # Load sample submission
    df = pd.read_csv(sample_submission_path)
    
    # Process each row
    results = []
    for idx, row in df.iterrows():
        original_text = row['text']
        
        # Initialize and run Hybrid Optimizer
        optimizer = HybridPerplexityPermutationOptimizer(
            scorer=scorer, 
            original_text=original_text,
            early_stopping=True
        )
        
        best_permutation, best_perplexity = optimizer.optimize()
        
        # Store result
        results.append({
            'id': idx,
            'text': ' '.join(best_permutation)
        })
        
        print(f"Row {idx}: Best Perplexity = {best_perplexity:.4f}")
        
        # Optional: Free GPU memory
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
    
    # Create submission DataFrame
    submission_df = pd.DataFrame(results)
    submission_df.to_csv(output_path, index=False)
    print(f"Submission saved to {output_path}")
    
    return submission_df

In [5]:
# Example usage
solve_santa_2024_permutation_SA('/kaggle/input/santa-2024/sample_submission.csv')

Loading checkpoint shards:   0%|          | 0/8 [00:00<?, ?it/s]

Starting from v4.46, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)


Iteration 1/200: Best Perplexity = 556.4528
Iteration 2/200: Best Perplexity = 556.4528
Iteration 3/200: Best Perplexity = 556.4528
Iteration 4/200: Best Perplexity = 556.4528
Iteration 5/200: Best Perplexity = 556.4528
Iteration 6/200: Best Perplexity = 556.4528
Iteration 7/200: Best Perplexity = 556.4528
Iteration 8/200: Best Perplexity = 556.4528
Iteration 9/200: Best Perplexity = 556.4528
Iteration 10/200: Best Perplexity = 556.4528
Iteration 11/200: Best Perplexity = 556.4528
Early stopping triggered.
Row 0: Best Perplexity = 556.4528
Iteration 1/200: Best Perplexity = 852.1607
Iteration 2/200: Best Perplexity = 852.1607
Iteration 3/200: Best Perplexity = 852.1607
Iteration 4/200: Best Perplexity = 852.1607
Iteration 5/200: Best Perplexity = 852.1607
Iteration 6/200: Best Perplexity = 852.1607
Iteration 7/200: Best Perplexity = 852.1607
Iteration 8/200: Best Perplexity = 852.1607
Iteration 9/200: Best Perplexity = 852.1607
Iteration 10/200: Best Perplexity = 852.1607
Iteration 11/

id                                               text
0   0  reindeer mistletoe scrooge gingerbread chimney...
1   1  reindeer the elf scrooge family sleep walk nig...
2   2  sleigh yuletide naughty nice holiday cheer gri...
3   3  magi grinch holiday yuletide cheer relax unwra...
4   4  merry hohoho it snowglobe toy game puzzle from...
5   5  and hope joy the grinch relax eat unwrap give ...

### Results:

- **Row 0**: 556.4528
- **Row 1**: 852.1607
- **Row 2**: 500.7958
- **Row 3**: 442.9493
- **Row 4**: 501.2758
- **Row 5**: 322.2915

The mean perplexity across all rows is approximately **529.32**. This indicates an overall improvement in the perplexity as the hybrid optimization algorithm progresses. The decrease in perplexity values across the rows, especially from Row 0 (556.45) to Row 5 (322.29), suggests that the hybrid approach effectively reduces perplexity, which is commonly associated with better model performance, likely leading to improved optimization and solution quality in the context of the problem being solved.
